MSAL/src/cache/MSALTokenCache.m

```diff
@@ -144,107 +144,111 @@ - (BOOL)saveAccessToken:(MSALAccessTokenCacheItem *)atItem
+                               error:(NSError * __autoreleasing *)error
 {
     [[MSALTelemetry sharedInstance] startEvent:[ctx telemetryRequestId] eventName:MSAL_TELEMETRY_EVENT_TOKEN_CACHE_LOOKUP];
     MSALTelemetryCacheEvent *event = [[MSALTelemetryCacheEvent alloc] initWithName:MSAL_TELEMETRY_EVENT_TOKEN_CACHE_LOOKUP
                                                                            context:ctx];
     [event setTokenType:MSAL_TELEMETRY_VALUE_ACCESS_TOKEN];
     
-    NSArray<MSALAccessTokenCacheItem *> *allAccessTokens = [self allAccessTokensForUser:user
-                                                                               clientId:clientId
-                                                                                context:ctx
-                                                                                  error:error];
-    if (!allAccessTokens)
+    BOOL ret =  [self findAccessTokenImpl:authority
+                                 clientId:clientId
+                                   scopes:scopes
+                                     user:user
+                                  context:ctx
+                              accessToken:outAccessToken
+                           authorityFound:outAuthorityFound
+                                    error:error];
+    
+    [[MSALTelemetry sharedInstance] stopEvent:[ctx telemetryRequestId] event:event];
+    
+    return ret;
+}
+
+- (BOOL)findAccessTokenImpl:(NSURL *)authority
+                   clientId:(NSString *)clientId
+                     scopes:(MSALScopes *)scopes
+                       user:(MSALUser *)user
+                    context:(nullable id<MSALRequestContext>)ctx
+                accessToken:(MSALAccessTokenCacheItem **)outAccessToken
+             authorityFound:(NSString **)outAuthorityFound
+                      error:(NSError * __autoreleasing *)error
+{
+    REQUIRED_PARAMETER_BOOL(user, ctx);
+    REQUIRED_PARAMETER_BOOL(outAccessToken, ctx);
+    REQUIRED_PARAMETER_BOOL(outAuthorityFound, ctx);
+    
+    *outAccessToken = nil;
+    *outAuthorityFound = nil;
+    
+    NSArray<MSALAccessTokenCacheItem *> *allAccessTokens =
+    [self allAccessTokensForUser:user
+                        clientId:clientId
+                         context:ctx
+                           error:error];
+    if (!allAccessTokens || allAccessTokens.count == 0)
     {
-        [[MSALTelemetry sharedInstance] stopEvent:[ctx telemetryRequestId] event:event];
+        // This should be rare-to-never as having a MSALUser object requires having a RT in cache,
+        // which should imply that at some point we got an AT for that user with this client ID
+        // as well. Unless users start working cross client id of course.
+        LOG_WARN(ctx, @"No access token found for user & client id.");
+        LOG_WARN_PII(ctx, @"No access token found for user & client id.");
         
-        return nil;
+        return NO;
```

Comment:
This is special cased in MSALSilentRequest:

```
if (error == nil && !_parameters.unvalidatedAuthority)
```

In the case where we didn't find any ATs at all in the cache, and the developer did not provide an authority string to use then under our current spec'd behavior we don't have an authority to fall back on and have to fail out.